contracts/REXTwoWayMaticMarket.sol

```diff
@@ -0,0 +1,463 @@
+  ) external view virtual override returns (bytes memory _cbdata) {
+      _onlyHost();
+      _onlyExpected(_superToken, _agreementClass);
+
+      (address _shareholder, int96 _flowRateMain, uint256 _timestamp) = _getShareholderInfo(_agreementData, _superToken);
+
+      uint256 _uinvestAmount = _calcUserUninvested(
+          _timestamp,
+          uint256(uint96(_flowRateMain)),
+          // Select the correct lastDistributionAt for this _superToken
+          _getLastDistributionAt(_superToken)
+      );
+      _cbdata = abi.encode(_uinvestAmount, int(_flowRateMain));
+
+  }
+
+
+  function afterAgreementTerminated(
+      ISuperToken _superToken,
+      address _agreementClass,
+      bytes32, //_agreementId,
+      bytes calldata _agreementData,
+      bytes calldata _cbdata, //_cbdata,
+      bytes calldata _ctx
+  ) external virtual override returns (bytes memory _newCtx) {
+      _onlyHost();
+      _onlyExpected(_superToken, _agreementClass);
+
+      _newCtx = _ctx;
+      (address _shareholder, ) = abi.decode(_agreementData, (address, address));
+      (uint256 _uninvestAmount, int96 _beforeFlowRate ) = abi.decode(_cbdata, (uint256, int96));
+
+      ShareholderUpdate memory _shareholderUpdate = ShareholderUpdate(
+        _shareholder, _beforeFlowRate, 0, _superToken
+      );
+
+      _newCtx = _updateShareholder(_newCtx, _shareholderUpdate);
+      // Refund the unswapped amount back to the person who started the stream
+      _superToken.transferFrom(
+          address(this),
+          _shareholder,
+          _uninvestAmount
+      );
+  }
+
+
+  function _swap(
+        ISuperToken input,
+        ISuperToken output,
+        uint256 amount
+  ) internal {
+
+   address inputToken;           // The underlying input token address
+   address outputToken;          // The underlying output token address
+   address[] memory path;        // The path to take
+   uint256 minOutput;            // The minimum amount of output tokens based on Tellor
+   uint256 outputAmount;         // The balance before the swap
+
+   inputToken = input.getUnderlyingToken();
+   outputToken = output.getUnderlyingToken();
+
+   // Downgrade and scale the input amount
+   input.downgrade(amount);
+   // Scale it to 1e18 for calculations
+   amount = ERC20(inputToken).balanceOf(address(this)) * (10 ** (18 - ERC20(inputToken).decimals()));
+
+   minOutput = amount * market.oracles[input].usdPrice / market.oracles[output].usdPrice;
+   minOutput = minOutput * (1e6 - market.rateTolerance) / 1e6;
+
+   // Scale back from 1e18 to outputToken decimals
+   minOutput = minOutput * (10 ** (ERC20(outputToken).decimals())) / 1e18;
+   // Scale it back to inputToken decimals
+   amount = amount / (10 ** (18 - ERC20(inputToken).decimals()));
+
+   // Assumes a direct path to swap input/output
+   path = new address[](2);
+   path[0] = inputToken;
+   path[1] = outputToken;
+
+   // Convert the outputToken back to its supertoken version
```

Comment:
Below is the start of the main change for this contract from the `REXTwoWayMarket.sol` contract